internal/controller/bucket/policy.go

```diff
@@ -0,0 +1,203 @@
+	"github.com/linode/provider-ceph/internal/rgw"
+	"go.opentelemetry.io/otel"
+)
+
+// PolicyClient is the client for API methods and reconciling the BucketPolicy
+type BucketPolicyClient struct {
+	backendStore    *backendstore.BackendStore
+	s3ClientHandler *s3clienthandler.Handler
+	log             logging.Logger
+}
+
+// NewBucketPolicyClient creates the client for Accelerate Configuration
+func NewBucketPolicyClient(b *backendstore.BackendStore, h *s3clienthandler.Handler, l logging.Logger) *BucketPolicyClient {
+	return &BucketPolicyClient{backendStore: b, s3ClientHandler: h, log: l}
+}
+
+//nolint:dupl // LifecycleConfiguration and BucketPolicy are different feature.
+func (p *BucketPolicyClient) Observe(ctx context.Context, bucket *v1alpha1.Bucket, backendNames []string) (ResourceStatus, error) {
+	ctx, span := otel.Tracer("").Start(ctx, "bucket.BucketPolicyClient.Observe")
+	defer span.End()
+
+	observationChan := make(chan ResourceStatus)
+	errChan := make(chan error)
+
+	for _, backendName := range backendNames {
+		beName := backendName
+		go func() {
+			observation, err := p.observeBackend(ctx, bucket, beName)
+			if err != nil {
+				errChan <- err
+
+				return
+			}
+			observationChan <- observation
+		}()
+	}
+
+	for i := 0; i < len(backendNames); i++ {
+		select {
+		case <-ctx.Done():
+			p.log.Info("Context timeout during bucket policy observation", consts.KeyBucketName, bucket.Name)
+			err := errors.Wrap(ctx.Err(), errObservePolicy)
+			traces.SetAndRecordError(span, err)
+
+			return NeedsUpdate, err
+		case observation := <-observationChan:
+			if observation != Updated {
+				return observation, nil
+			}
+		case err := <-errChan:
+			err = errors.Wrap(err, errObservePolicy)
+			traces.SetAndRecordError(span, err)
+
+			return NeedsUpdate, err
+		}
+	}
+
+	return Updated, nil
+}
+
+func (p *BucketPolicyClient) observeBackend(ctx context.Context, bucket *v1alpha1.Bucket, backendName string) (ResourceStatus, error) {
+	p.log.Info("Observing subresource policy on backend", consts.KeyBucketName, bucket.Name, consts.KeyBackendName, backendName)
+
+	if p.backendStore.GetBackendHealthStatus(backendName) == apisv1alpha1.HealthStatusUnhealthy {
+		// If a backend is marked as unhealthy, we can ignore it for now by returning Updated.
+		// The backend may be down for some time and we do not want to block Create/Update/Delete
+		// calls on other backends. By returning NeedsUpdate here, we would never pass the Observe
+		// phase until the backend becomes Healthy or Disabled.
+		return Updated, nil
+	}
+
+	s3Client, err := p.s3ClientHandler.GetS3Client(ctx, bucket, backendName)
+	if err != nil {
+		return NeedsUpdate, err
+	}
+
+	// external keeps the bucket policy in backend.
+	var external string
+
+	response, err := rgw.GetBucketPolicy(ctx, s3Client, aws.String(bucket.Name))
+	// If error is not NoSuchBucketPolicy error, return with the error.
+	if err != nil && !isNoSuchBucketPolicy(err) {
```

Comment:
If you haven't already, we should validate this error against an actual Ceph cluster. There can be inconsistencies between S3 and Ceph error messages in my experience.

See [here](https://github.com/linode/provider-ceph/blob/main/docs/DEVELOPMENT.md#running-locally) for running locally against a remote Ceph cluster.

Reply:
I confirmed that Ceph also returns NoSuchBucketPolicy error and isNoSuchBucketPolicy function works appropriately.